language/move-prover/spec-lang/src/env.rs

```diff
@@ -1906,6 +1974,37 @@ impl<'env> FunctionEnv<'env> {
             .any(|Parameter(_, ty)| ty.is_mutable_reference())
     }
 
+    /// Returns the types of mutable reference function arugments
+    pub fn get_mut_ref_types(&self) -> Vec<Type> {
+        self.get_parameters()
+            .into_iter()
+            .filter(|Parameter(_, ty)| ty.is_mutable_reference())
+            .map(|Parameter(_, ty)| ty.clone())
+            .collect_vec()
+    }
+
+    /// Returns the number of mutable references
+    pub fn get_mut_ref_count(&self) -> usize {
+        self.get_mut_ref_types().len()
+    }
+
+    /// Returns the types of the return values of the procedure.
+    /// This includes the values returned by the function and the
+    /// mutable reference arguments.
+    pub fn get_proc_return_types(&self) -> Vec<Type> {
```

Comment:
So this set of functions looks like it is simulating the effect of the memory model transformation. As such it does not belong into FunctionEnv but in FunctionTarget. In fact I don't remember how Shaz has been doing this, but really the FunctionTarget::get_parameters and FunctionTarget::get_return_type already should do the right thing here. @shazqadeer for confirmation how to deal with this.